src/Processors/Executors/PullingPipelineExecutor.cpp

```diff
@@ -0,0 +1,186 @@
+#include <Processors/Executors/PullingPipelineExecutor.h>
+#include <Processors/Executors/PipelineExecutor.h>
+#include <Processors/Formats/LazyOutputFormat.h>
+#include <Processors/Transforms/AggregatingTransform.h>
+#include <Processors/QueryPipeline.h>
+
+#include <Common/setThreadName.h>
+#include <ext/scope_guard.h>
+
+namespace DB
+{
+
+struct PullingPipelineExecutor::Data
+{
+    PipelineExecutorPtr executor;
+    std::exception_ptr exception;
+    std::atomic_bool is_executed = false;
+    std::atomic_bool has_exception = false;
+    ThreadFromGlobalPool thread;
+
+    ~Data()
+    {
+        if (thread.joinable())
+            thread.join();
+    }
+};
+
+PullingPipelineExecutor::PullingPipelineExecutor(QueryPipeline & pipeline_) : pipeline(pipeline_)
+{
+    lazy_format = std::make_shared<LazyOutputFormat>(pipeline.getHeader());
+    pipeline.setOutput(lazy_format);
+}
+
+PullingPipelineExecutor::~PullingPipelineExecutor()
+{
+    try
+    {
+        /// Cancel execution if it wasn't finished.
+        if (data && !data->is_executed)
+            cancel();
+
+        /// Finish lazy format. Otherwise thread.join() may hung.
+        if (!lazy_format->isFinished())
+            lazy_format->finish();
+
+        /// Join thread here to wait for possible exception.
+        if (data && data->thread.joinable())
+            data->thread.join();
+    }
+    catch (...)
+    {
+        tryLogCurrentException("PullingPipelineExecutor");
+    }
+
+    if (data && data->has_exception)
+        tryLogException(std::move(data->exception), "PullingPipelineExecutor");
+}
+
+static void threadFunction(PullingPipelineExecutor::Data & data, ThreadGroupStatusPtr thread_group, size_t num_threads)
```

Comment:
Why this function is not a private member of PullingPipelineExecutor?

Reply:
It may be, but I prefer to hide implementation details from header file.
Don't think that is is very important.